samples/gamma-fade/main.c

```diff
@@ -0,0 +1,255 @@
+#include <hal/debug.h>
+#include <windows.h>
+
+static float remap(float in_low,  float in_high,
+                   float out_low, float out_high,
+                   float value)
+{
+                                 /* [in_low; in_high]       */
+    value -= in_low;             /* [0.0; in_high-in_low]   */
+    value /= in_high - in_low;   /* [0.0; 1.0]              */
+    value *= out_high - out_low; /* [0.0; out_high-out_low] */
+    value += out_low;            /* [out_low; out_high]     */
+    return value;
+}
+
+/* This is also known as `lerp(a, b, t)` */
+static float mix(float a, float b, float t)
+{
+    return a * (1.0f - t) + b * t;
+}
+
+/* This is also known as `mix(0.0, 1.0, x)` */
+static float linear(float x)
+{
+    return x;
+}
+
+/* This is also known as `smoothstep(0.0, 1.0, x)` */
+static float smooth(float x)
+{
+    return x * x * (3.0f - 2.0f * x);
+}
+
+static void set_mix_fade(float red, float green, float blue,
+                         float t)
+{
+    GAMMA_RAMP_ENTRY entries[256];
+    for(int i = 0; i < 256; i++) {
+        float f = i / (float)0xFF;
+
+        entries[i].red   = 0xFF * mix(f, red,   t);
+        entries[i].green = 0xFF * mix(f, green, t);
+        entries[i].blue  = 0xFF * mix(f, blue,  t);
+    }
+    XVideoSetGammaRamp(0, entries, 256);
+}
+
+static void set_cinematic_fade(float a_black_input,  float a_white_input,
+                               float a_black_output, float a_white_output,
+                               float b_black_input,  float b_white_input,
+                               float b_black_output, float b_white_output,
+                               float t)
+{
+    GAMMA_RAMP_ENTRY entries[256];
+    for(int i = 0; i < 256; i++) {
+        float f = i / (float)0xFF;
+
+        /* Interpolate the input and output levels */
+        float black_input  = mix(a_black_input,  b_black_input,  t);
+        float white_input  = mix(a_white_input,  b_white_input,  t);
+        float black_output = mix(a_black_output, b_black_output, t);
+        float white_output = mix(a_white_output, b_white_output, t);
+
+        /* Remap our brightness */
+        if (f < black_input) { f = black_input; }
+        if (f > white_input) { f = white_input; }
+        f = remap(black_input, white_input, black_output, white_output, f);
+        if (f < black_output) { f = black_output; }
+        if (f > white_output) { f = white_output; }
+
+        entries[i].red   = 0xFF * f;
+        entries[i].green = 0xFF * f;
+        entries[i].blue  = 0xFF * f;
+    }
+    XVideoSetGammaRamp(0, entries, 256);
+}
+
+static float animate_frame(DWORD ts_start, DWORD ts, float duration)
+{
+    /* Wait for the next frame */
+    XVideoWaitForVBlank();
+
+    /* Update timestamp */
+    DWORD ts_delta = ts - ts_start;
+    float t = remap(0.0f, duration, 0.0f, 1.0f, ts_delta);
+
+    /* Signal completion */
+    if (t >= 1.0f) {
+        return 1.0f;
+    }
+
+    /* Return time in range [0.0; 1.0[ */
```

Comment:
Typo here

Reply:
I don't see it?

In case you mean the closing "[": The range has 1.0 excluded (it was handled above). [See this reference](https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints ).

Reply:
Oh I see what you mean here. In this case I would have used: `[0, 1)` but it appears `[0, 1[` is also some convention..